### PR TITLE
feat(auth): add getCurrentUser API

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes

The `getCurrentUser` API returns the current authenticated user 

**API changes**
- calls fetchAuthSession to get the `idToken` 
- accepts a `recache` flag that will refresh tokens and recache them in the storage mechanism

**Test  cases**
- tests that `getCurrentUser` returns the right `AuthUser`
- test that `fetchAuthSession` throws when trying to recache tokens

#### Description of how you validated changes
I tested this API with a sample App

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [x] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by HuiSF — DISMISSED** [review]

**Review by erinleigh90 — APPROVED** [review]

File: packages/auth/__tests__/providers/cognito/getCurrentUser.test.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+Amplify.configure({
+	Auth: {
+		userPoolWebClientId: '111111-aaaaa-42d8-891d-ee81a1549398',
+		userPoolId: 'us-west-2_zzzzz',
+		identityPoolId: 'us-west-2:xxxxxx',
+	},
+});
+const mockedAccessToken =
+	'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c';
```
Comment (stocaaro) [inline packages/auth/__tests__/providers/cognito/getCurrentUser.test.ts:22]: This looks like a real thing. Does it need to look like a real thing? Looking like a real thing is one of the way we can end up with security warnings.

File: packages/auth/__tests__/providers/cognito/getCurrentUser.test.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+							idToken: {
+								payload: {
+									sub: mockedSub,
+									'cognito:username': mockedUsername,
+								},
+							},
+						},
+					};
+				}
+			);
```
Comment (stocaaro) [inline packages/auth/__tests__/providers/cognito/getCurrentUser.test.ts:45]: nit: Seem like we're getting a bit deep in nested structures here. Probably fine for testing, but makes me wonder if there's not a better way to represent this return type or the layers of how it is returning.

**Review by stocaaro — APPROVED** [review]

File: packages/auth/src/providers/cognito/apis/getCurrentUser.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Gets the current user from the idToken.
+ *
+ * @param getCurrentUserRequest - The request object.
+ *
+ * @throws - {@link InitiateAuthException} - Thrown when the service fails to refresh the tokens.
+ *
+ * @throws AuthTokenConfigException - Thrown when the token provider config is invalid.
+ *
+ * @returns AuthUser
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/apis/getCurrentUser.ts:20]: nit

```suggestion
 * @returns A Promise that will resolve with an AuthUser.
```

**Review by jimblanc — APPROVED** [review]

File: packages/auth/__tests__/providers/cognito/getCurrentUser.test.ts (hunk 1 above)
Comment (israx) [reply packages/auth/__tests__/providers/cognito/getCurrentUser.test.ts:22]: Yeah I think it is not necessary. I can create another PR to change this implementation

File: packages/auth/src/providers/cognito/apis/getCurrentUser.ts (hunk 3 above)
Comment (israx) [reply packages/auth/src/providers/cognito/apis/getCurrentUser.ts:20]: Will address it in a next PR

**Pull request merged**